common/http/http.cpp

```diff
@@ -68,91 +79,154 @@ string MethodToString(Method method) {
 	return "INVALID_METHOD";
 }
 
-Request::Request(Method method) :
-	method_(method) {
-}
-
-expected::ExpectedString Transaction::GetHeader(const string &name) const {
-	if (headers_.find(name) == headers_.end()) {
-		return expected::unexpected(MakeError(NoSuchHeaderError, "No such header: " + name));
-	}
-	return headers_.at(name);
-}
-
-void Request::SetHeader(const string &name, const string &value) {
-	headers_[name] = value;
-}
-
-error::Error Request::SetAddress(const string &address) {
-	ready_ = false;
-
+error::Error BreakDownUrl(const string &url, BrokenDownUrl &address) {
 	const string url_split {"://"};
 
-	auto split_index = address.find(url_split);
+	auto split_index = url.find(url_split);
 	if (split_index == string::npos) {
-		return MakeError(InvalidUrlError, address + " is not a valid URL.");
+		return MakeError(InvalidUrlError, url + " is not a valid URL.");
 	}
 	if (split_index == 0) {
-		return MakeError(InvalidUrlError, address + ": missing protocol");
+		return MakeError(InvalidUrlError, url + ": missing hostname");
 	}
 
-	protocol_ = address.substr(0, split_index);
+	address.protocol = url.substr(0, split_index);
 
-	auto tmp = address.substr(split_index + url_split.size());
+	auto tmp = url.substr(split_index + url_split.size());
 	split_index = tmp.find("/");
 	if (split_index == string::npos) {
-		host_ = tmp;
-		path_ = "/";
+		address.host = tmp;
+		address.path = "/";
 	} else {
-		host_ = tmp.substr(0, split_index);
-		path_ = tmp.substr(split_index);
+		address.host = tmp.substr(0, split_index);
+		address.path = tmp.substr(split_index);
 	}
 
-	split_index = host_.find(":");
+	split_index = address.host.find(":");
```

Comment:
Good observation, but it's intentional, because the parser itself does not know which protocols are supported. There will be an error, but it's caught at a higher level, [here](https://github.com/mendersoftware/mender/pull/1189/files#diff-7100548573dbc4eefb8ff288d50eb33b3c038a36e5573de55fa670b1f0221c46R110).

Still good that you pointed it out, because I had forgotten to add the same check in `AsyncServeUrl`.